spec/app/ics-028-cross-chain-valset-update/README.md

```diff
@@ -0,0 +1,257 @@
+
+### Application data
+
+#### Parent blockchain
+
+- Q: The causally consistent FIFO shared queue.
+
+#### Baby blockchain
+
+- Q: The causally consistent FIFO shared queue.
+- *UnbondingQueue*: Keeps track of validators that are currently unbonding.
+- validatorSetSeqNum: Number of validator set change demands processed; initialized to 0.
+- dequeueSeqNum: Number of elements dequeued from Q; initialized to 0.
+- lastObserved: Element of Q that was observed during the last Q.read() operation; initializated to nil.
+
+## Transitions
+
+In this section, we informally discuss the state transitions that occur in our protocol.
+We observe state transitions that are driven by a user (i.e., the new staking module on the parent chain), driven by the relayer and driven by elapsed time.
+
+  - User-driven state transitions: These transitions "start" the entire process of changing the validator set of the baby blockchain.
+  We assume that the will to change the validator set of the baby blockchain babyChain is expressed by invoking <ChangeValidatorSet, babyChain, V>.
+
+  - Time-driven state transitions: These transitions are activated since some time has elapsed. As we will present in the rest of the document, time-driven state transitions help us determine when the unbonding period, measured at the baby blockchain, has elapsed for a validator set.
+
+In the rest of the document, we will discuss the aforementioned state transitions in more detail.
+
+## Function Definitions
+
+### Parent blockchain
+
+This subsection will present the functions executed at the parent blockchain.
+
+```golang
+// expresses will to modify the validator set of the baby blockchain;
+func changeValidatorSet(
+  babyChainId: ChainId
+  valSetUpdate: Validator[]
+) {
+  // enqueue to the queue shared with babyChainId
+  Q[babyChainId].enqueue(valSetUpdate)
+}
+```
+
+- Expected precondition
+  - There exists a blockchain with *babyChainId* identifier
+  - All validators from *valSetUpdate* are validators at the parent blockchain
+- Expected postcondition
+  - The valSetUpdate is enqueued to the queue shared with the blockchain with *babyChainId*
+- Error condition
+  - If the precondition is violated
+
+TODO discuss and write a staking logic on endBlock
+Important: By reading the shared queue Q, the parent blockchain is able to conclude which validators could get their stake back.
+
+### Baby blockchain
+
+```golang
+// initialization function
+func init() {
+  UnbondingQueue = new local FIFO queue
+  validatorSetSeqNum = 0
+  dequeueSeqNum = 0
+  lastObserved = nil
+}
+```
+
+
+```golang
+// function used in End-Block method
+// returns all the elements enqueued since the last invocation of this method
+func observeChangesQ() { 
+  // get the content of the shared queue
+  content = Q.read()
+
+  // check whether the last observed element is in content
```

Comment:
Do we need to do all of this? Can't we just iterate over anything in the queue and apply the changes?

Reply:
We need to figure out which updates are "new". Otherwise, we could apply one validator set update more than once, which we want to avoid.

Reply:
I think this question resulted from my confusion in how the queue was being used, since clarified.